src/core/mainloop/mainloop.c

```diff
@@ -2003,6 +2005,55 @@ add_entropy_callback(time_t now, const or_options_t *options)
   return ENTROPY_INTERVAL;
 }
 
+/** Periodic callback: if there has been no network usage in a while,
+ * enter a dormant state. */
+static int
+check_network_participation_callback(time_t now, const or_options_t *options)
+{
+  /* If we're a server, we can't become dormant. */
+  if (server_mode(options)) {
+    note_user_activity(now);
+    goto end;
+  }
+
+  /* If we're running an onion service, we can't become dormant. */
+  /* XXXX this would be nice to change, so that we can be dormant with a
+   * service. */
+  if (hs_service_get_num_services() || rend_num_services()) {
+    note_user_activity(now);
+    goto end;
+  }
+
+  /* XXXX Add an option to never become dormant. */
+
+  /* If we have any currently open entry streams other than "linked"
+   * connections used for directory requests, those count as user activity.
+   */
+  /* XXXX make this configurable? */
```

Comment:
Not sure... I think that is the sane default and anything that wants to have more fine grained control should use a signal or control port to tell tor what to do.

Maybe when the use cases for more fine grained control here arise, we can add an option but for now I'm ok with this default hardcoded.